### ISSUE TITLE
SynthDef.writeOnce is a developer trap

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#1910]
The docs say that SynthDef.writeOnce is good for class definitions to avoid rewriting the SynthDef every time the class library is compiled. But it's not actually useful in its intended use case, because it's far too resilient to change.

Suppose I release version 1.0 of a quark, and it has a SynthDef.writeOnce in it. Someone downloads my quark for the first time and my SynthDef is written once to their filesystem. If I release version 1.1 of the quark with a modified SynthDef, someone who upgrades will have the SynthDef from version 1.0 and the rest of the library from 1.1. Difficult-to-diagnose bugs ensue.

The method should be removed, or changed so that the behavior is not conducive to bugs by default.

List discussion is here: http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/writeOnce-td7624131.html

**Comment by jamshark70** [comment]
An important change in the last couple of years is that all defs in the global SynthDescLib are sent to any server when it's booted. There is no longer any need to boot the server first and then .add a def -- you can .add anytime, and it will be available after boot.

Hence there is no longer any need to use writeOnce at all.

So my suggested solution is simply: don't use writeOnce. And document that this is a legacy method to be generally avoided.

**Comment by jamshark70** [comment]
Correction: There is no longer any need to use writeOnce *apart from very large SynthDefs*.

For the large def case, you could use metadata to store the def with a version number. Then, during class init, check the metadata first. If the version number doesn't match, then and only then rebuild.

I have to admit, massive defs are a valid case, as it can take half a second or more to optimize and topo-sort the really huge ones.

**Comment by snappizz** [comment]
It sounds like we should introduce a new method (.cache?) specifically intended for very large SynthDefs. Then we deprecate .writeOnce and recommend .add for small SynthDefs and .cache for large ones.

**Comment by jamshark70** [comment]
Interesting factoid: `writeOnce` always builds the SynthDef! IMO the only reason to use `writeOnce` is to avoid the overhead of building a def that is already there... but it calls `SynthDef.new` which goes through the entire building process.

I'd raise the same objection to the hash approach that was proposed on the mailing list. You have to build the def to calculate the hash, losing the optimization entirely.

Here's a proof-of-concept of the metadata approach:

```
(
var desc;

SynthDescLib.read;
desc = SynthDescLib.global[\test];

// def not found: need to create
if(desc.isNil or: {
	// missing metadata, can't verify version: re-create
	desc.metadata.isNil or: {
		// wrong version: re-create
		desc.metadata[\version] != "1.1"
	}
}) {
	"Writing def".postln;
	SynthDef(\test, {
		Out.ar(0, DC.ar(0))
	}, metadata: (version: "1.1")).store;
} {
	"def was already there".postln;
};
)
```

**Comment by crucialfelix** [comment]
.writeDefFile and .writeOnce are different and distinct to .add

They write to a file, and there are many things you might do with that file.
You could ship it with a standalone product.
You could load it on demand (from a large folder of them) from a separate
client.

Writing def files is core functionality and should be left alone.

If you reboot the language then everything added with .add is gone.
.writeDefFile can and is used by many people to pre-generated defs and then
load them on demand.

You can recommend that .add is better, but please don't label it as
"legacy" or "deprecated"




On Sun, Mar 20, 2016 at 2:57 AM jamshark70 <notifications@github.com> wrote:

> Interesting factoid: writeOnce always builds the SynthDef! IMO the only
> reason to use writeOnce is to avoid the overhead of building a def that
> is already there
>

that's something you could optimize in the implementation.
though since you don't use the method I wouldn't recommend bothering doing
it.


> ... but it calls SynthDef.new which goes through the entire building
> process.
>
> I'd raise the same objection to the hash approach that was proposed on the
> mailing list. You have to build the def to calculate the hash, losing the
> optimization entirely.
>
> Here's a proof-of-concept of the metadata approach:
>
> (
> var desc;
>
> SynthDescLib.read;
> desc = SynthDescLib.global[\test];
>
> // def not found: need to create
> if(desc.isNil or: {
>     // missing metadata, can't verify version: re-create
>     desc.metadata.isNil or: {
>         // wrong version: re-create
>         desc.metadata[\version] != "1.1"
>     }
> }) {
>     "Writing def".postln;
>     SynthDef(\test, {
>         Out.ar(0, DC.ar(0))
>     }, metadata: (version: "1.1")).store;
> } {
>     "def was already there".postln;
> };
> )
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/supercollider/supercollider/issues/1910#issuecomment-198826204>
>
-- 
https://twitter.com/crucialfelix
https://medium.com/@crucialfelix
http://www.mattermind.com/

**Comment by telephon** [comment]
>Interesting factoid: writeOnce always builds the SynthDef! IMO the only reason to use writeOnce is to avoid the overhead of building a def that is already there... but it calls SynthDef.new which goes through the entire building process.

Yes, it just omits the writing to disk. I would say, we should just leave it alone and not use it anymore in class inits. Is this a documentation issue then?

**Comment by snappizz** [comment]
I can't think of any cases where the current behavior of .writeOnce is useful, especially considering James' comment demonstrating it doesn't actually have all the advertised efficiency. By deprecating it, we do everyone a favor by alerting quark developers to the use of a language feature that is practically guaranteed to cause bugs.

I guess it's not so bad to just document it as a legacy feature and fix the quarks that use it, but my point is that we don't lose anything by removing it (aside from backward compatibility, which is a valid concern).

@crucialfelix I'm not sure I understand -- this discussion is about .writeOnce, not .writeDefFile. Are you saying that .writeOnce is core functionality and shouldn't be tampered with?

**Comment by crucialfelix** [comment]
I'm saying that they are not equivalent to or replaced by .add 
I mention them both because both of them are similar (writing files to disk)

Changing documentation is fine and a good idea. But deprecating means removing.
That means breaking supercollider code, and its a very commonly used method.
People are on the user list right now talking about using that method. 
So its not polite to just remove it.
That means guaranteed breakage of known user code.

**Comment by telephon** [comment]
It *does* have a use: When you want to avoid writing to disk more than absolutely necessary, it does exactly what it should. The method `writeDef` is not called if a file exists, and also of course the file IO is not done.

My suggestion is: post a brief message to the post window that it has not written a certain synthdef because it existed already. And add a waring to the helpfile that tells you about the potential risk.

**Comment by snappizz** [comment]
@telephon Sounds like a reasonable proposal. One objection: the post window messages could get annoying for the user if there are a lot of .writeOnce's in their class library.

**Comment by jamshark70** [comment]
Let's tread carefully. Tim  Walters, on the mailing list, raised a related but somewhat conflicting use case: very large temporary SynthDefs that should be written to disk for transmission, and then discarded -- and which should not necessarily be added to the global SynthDescLib. Currently SC handles this in a clumsy way too: the user is responsible for removing the SynthDef file, and a large def in the SynthDescLib adds a ton of objects in memory.

Maybe we should collect use cases before making decisions about object interfaces. Otherwise, we add a hack for this issue, and we add another hack for Tim's use case, and another etc. until nobody knows what is the right way to do anything.

**Comment by snappizz** [comment]
In the particular case of a small SynthDef in a class library, which is preferable -- .writeDefFile or .add?

**Comment by jamshark70** [comment]
Others may disagree, but I feel strongly that classes should not write files on disk without the user's permission (or explicit instruction) unless it's absolutely necessary. For small SynthDefs, that means .add. Large SynthDefs that are necessary for a class's operation, I'd say, try to avoid doing it during class init (again, unless there's really no other way).

**Comment by snappizz** [comment]
Cool. I'll file a PR to fix the documentation.

**Comment by crucialfelix** [comment]
yes, I definitely agree all the classes should stop writing to disk. I find it really annoying that systems that I have never used and will never use insist on writing and loading defs. it should be lazy (on demand).

**Comment by snappizz** [comment]
Fixed in #1918. There's ongoing discussion on the more general problem of how to efficiently handle large SynthDefs in the class library, and I encourage others to file relevant issues.
